lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -813,12 +814,11 @@ defmodule Trento.Domain.SapSystem do
   defp maybe_emit_sap_system_deregistered_event(
          %SapSystem{
            sap_system_id: sap_system_id,
-           database: %Database{
-             sid: nil
-           }
+           database: %Database{deregistered_at: database_deregistered_at}
```

Comment:
I'm missing something here. We should check both, database `deregistered_at` and sap system `deregistered_at` . Something like:
```
when not is_nil(database_deregistered_at) or not is_nil(sapsystem_deregistered_at)
```
Or add this condition to the next `maybe_emit_sap_system_deregistered_event` clause. I'm not sure.

Otherwise we can have `SapSystemDeregistered` emitted twice, one when the database is deregistered, and 2nd when the applications are deregistered.

At the end, we need to check if the SAP system was deregistered somewhere to avoid emitting this event.

Reply:
Nice catch, adding a test and shipping the fix asap

Reply:
Done!

Reply:
The sap system is deregistered only if it's not already deregistered, following the database deregistered emit strategy, I just added the deregistered_at nil pattern match in function signatures and a black hole signature for capturing already deregistered sap systems with a no-op